selfdrive/ui/replay/replay.cc

```diff
@@ -1,8 +1,55 @@
 #include "replay.hpp"
 
-Replay::Replay(QString route_, int seek) : route(route_) {
-  unlogger = new Unlogger(&events, &events_lock, &frs, seek);
-  current_segment = 0;
+int getch(void) {
+  int ch;
+  struct termios oldt;
+  struct termios newt;
+
+  tcgetattr(STDIN_FILENO, &oldt);
+  newt = oldt;
+  newt.c_lflag &= ~(ICANON | ECHO);
+
+  tcsetattr(STDIN_FILENO, TCSANOW, &newt);
+  ch = getchar();
+  tcsetattr(STDIN_FILENO, TCSANOW, &oldt);
+
+  return ch;
+}
+
+Replay::Replay(QString route_, int seek_) : route(route_), seek(seek_) {
+  ctx = Context::create();
+  seek_request = seek*1e9;
+
+  QStringList block = QString(getenv("BLOCK")).split(",");
+  qDebug() << "blocklist" << block;
+
+  QStringList allow = QString(getenv("ALLOW")).split(",");
+  qDebug() << "allowlist" << allow;
+
+  for (const auto& it : services) {
+    std::string name = it.name;
+    if (allow[0].size() > 0 && !allow.contains(name.c_str())) {
+      qDebug() << "not allowing" << name.c_str();
+      continue;
+    }
+
+    if (block.contains(name.c_str())) {
+      qDebug() << "blocking" << name.c_str();
+      continue;
+    }
+
+    PubSocket *sock = PubSocket::create(ctx, name);
```

Comment:
instead of passing a SubMaster to process, pass it to the constructor. you don't want to make these sockets if we're in socketless mode